Optimize signed integer saturating_mul

xfix opened the pull request:
Previous version of saturating_mul was poorly optimized by LLVM, and had lots of branches (which are slow if CPU mispredicts them). That also had a side-effect of needlessly making code larger.

```asm
    mov     eax, edi
    imul    eax, esi
    jno     .LBB1_6
    mov     eax, -2147483648
    test    edi, edi
    js      .LBB1_4
    je      .LBB1_6
    test    esi, esi
    jg      .LBB1_5
    jmp     .LBB1_6
.LBB1_4:
    test    esi, esi
    jns     .LBB1_6
.LBB1_5:
    mov     eax, 2147483647
.LBB1_6:
    ret
```

However, it's possible to have branch-less saturating integer multiplication by using the fact that sign of exclusive xor is the same as infinite precision multiplication. This simplifies a complex number comparison function into a one simple test that can be easily implemented in assembly as a conditional move.

This assembly does also use the fact that minimal value is just one bigger than maximal value in signed representation. This wasn't done in the code, because LLVM can optimize that.

```asm
    mov     eax, esi
    xor     eax, edi
    shr     eax, 31
    add     eax, 2147483647
    imul    edi, esi
    cmovo   edi, eax
    mov     eax, edi
    ret
```

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @sfackler (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

sfackler commented:
This seems reasonable to me I think. cc @bluss 

@bors r+

bors commented:
:pushpin: Commit 888a2d7 has been approved by `sfackler`

<!-- @bors r=sfackler 888a2d78c86943f53b0a142cea5dfc1753315031 -->

bluss commented:
looks good

nagisa commented:
I would like to take a moment to note that this potentially pessimises the common case. Sure the original assembly looks intimidatingly big and branchy, compared to the version in the changed output, however if you take a closer look, you’ll notice that in the original case, where no overflow happens (i.e. what one would call the common case), there’s only 3 instructions for the compiler to execute (mov, multiply, branch), whereas in the new version 7 instructions are executed *unconditionally*.

Now, if the function ended looking like this instead:

```rust
a.checked_mul(other).unwrap_or_else(|| if (a ^ other).is_negative() {
    i64::min_value()
} else {
    i64::max_value()
})
```

you end up with

```asm
	mov	rax, rdi
	imul	rax, rsi
	jno	.LBB0_2
	xor	rsi, rdi
	shr	rsi, 63
	movabs	rax, 9223372036854775807
	add	rax, rsi
.LBB0_2:
	ret
```

which has exactly the same amount of branches and instructions as the original code, but still only executes 3 instructions for the common case (mov, multiply, branch).

---

In the end, I would like that landing such optimisations required some actual **numbers** showing the actual benefits of the change.

sfackler commented:
Yeah, I agree with @nagisa's request for numbers, sorry

@bors r-

xfix commented:
Sounds fair. I'm not sure how to benchmark code, so feel free to point out mistakes (benchmarking on CPUs today is tricky).

 I wrote a benchmark code which appears to compile to the code I want if I ask it to generate assembly source code (it doesn't optimize out saturating multiplication with just its result).

```rust
#![feature(test)]

extern crate test;

pub fn branchless(a: i32, b: i32) -> i32 {
    let overflow_value = if (a ^ b).is_negative() {
        i32::min_value()
    } else {
        i32::max_value()
    };

    a.checked_mul(b).unwrap_or(overflow_value)
}

pub fn branch(a: i32, b: i32) -> i32 {
    a.checked_mul(b).unwrap_or_else(|| if (a ^ b).is_negative() {
        i32::min_value()
    } else {
        i32::max_value()
    })
}

use test::Bencher;

#[bench]
fn bench_branchless(b: &mut Bencher) {
    b.iter(|| (0..1000).fold(0, branchless));
}

#[bench]
fn bench_branch(b: &mut Bencher) {
    b.iter(|| (0..1000).fold(0, branch));
}
```

This benchmark constantly multiplies numbers by 0, therefore avoiding any possible overflows.

For a benchmark, I changed scaling governor to `performance` (`powersave` made results really inconsistent).

```
running 2 tests
test bench_branch     ... bench:       2,538 ns/iter (+/- 12)
test bench_branchless ... bench:       2,568 ns/iter (+/- 2)
```

Code with branches is actually a bit faster.

When I changed `(0..1000).fold(0` to `(1..1000).fold(1` to make this trigger saturation arithmetic, results have changed.

```
test bench_branch     ... bench:       2,814 ns/iter (+/- 34)
test bench_branchless ... bench:       2,614 ns/iter (+/- 15)
```

So code with branches is a bit faster when overflow does not occur, while branchless code is faster when overflow does occur.

Tradeoffs, tradeoffs... I honestly don't know which one to go with anymore.

xfix commented:
Nevermind, my proposal is slower. Somehow the original code is the fastest one.

I suppose it makes sense, when a lot of comparisons trigger the same branch, not much of a code is executed. With 0 (no saturation branches).

```
test bench_branch     ... bench:       2,539 ns/iter (+/- 31)
test bench_branchless ... bench:       2,567 ns/iter (+/- 22)
test bench_orig       ... bench:       2,541 ns/iter (+/- 7)
```

With 1.

```
test bench_branch     ... bench:       2,638 ns/iter (+/- 30)
test bench_branchless ... bench:       2,666 ns/iter (+/- 403)
test bench_orig       ... bench:       2,338 ns/iter (+/- 84)
```

So I suppose I will close my pull request. At least I learned something about computers, so there is that.

xfix commented:
Although I suppose unit tests from this pull request could be added, but I don't think it's that important myself. If you like, I will submit just tests to this commit.

nagisa commented:
The benchmark is missing the black_boxes to ensure that the compiler wouldn’t optimise for the knowledge of input values, which probably means that you probably only checked the assembly “looked” right for only one or two versions you checked. 

[These](https://gist.github.com/nagisa/4463801bb64cf57b2848c2cd273be30e) are the results I could get. Here the branchy version (not the original one) is consistently the fastest, even if by very little. And it does compile to less code than the original.

Could you try running that benchmark on your machine? (Do not forget to use `-O` for optimisations!)

xfix commented:
I currently cannot run a benchmark, but it should be available in 5 hours. Please don't merge this right now, I just reopened this as a reminder to myself.

xfix commented:
Laptop (Sandybridge Pentium), Arch Linux:

```
running 9 tests
test bench_branch              ... bench:      25,194 ns/iter (+/- 339)
test bench_branch_oflo         ... bench:      25,197 ns/iter (+/- 273)
test bench_branch_oflo_neg     ... bench:      25,184 ns/iter (+/- 119)
test bench_branchless          ... bench:      26,201 ns/iter (+/- 92)
test bench_branchless_oflo     ... bench:      26,202 ns/iter (+/- 222)
test bench_branchless_oflo_neg ... bench:      26,203 ns/iter (+/- 255)
test bench_original            ... bench:      23,545 ns/iter (+/- 142)
test bench_original_oflo       ... bench:      23,550 ns/iter (+/- 357)
test bench_original_oflo_neg   ... bench:      25,794 ns/iter (+/- 62)
```

And because I realized that FreeBSD actually has Rust nightly package, I decided to benchmark it too.

VPS (Intel(R) Xeon(R) CPU E5-2650L v3 @ 1.80GHz), FreeBSD:

```
test bench_branch              ... bench:      21,657 ns/iter (+/- 1,530)
test bench_branch_oflo         ... bench:      21,939 ns/iter (+/- 1,753)
test bench_branch_oflo_neg     ... bench:      21,700 ns/iter (+/- 6,700)
test bench_branchless          ... bench:      26,311 ns/iter (+/- 1,888)
test bench_branchless_oflo     ... bench:      25,892 ns/iter (+/- 978)
test bench_branchless_oflo_neg ... bench:      25,692 ns/iter (+/- 678)
test bench_original            ... bench:      16,778 ns/iter (+/- 777)
test bench_original_oflo       ... bench:      19,541 ns/iter (+/- 1,322)
test bench_original_oflo_neg   ... bench:      16,682 ns/iter (+/- 2,342)
```

Just curious, what is your CPU? It may be because of my old CPU model.

Either way, whether the optimization works or not seems to be depending on exact CPU model.

xfix commented:
Committing branched version in event you want to push it, but I suppose this pull request should be just closed.

nagisa commented:
> Either way, whether the optimization works or not seems to be depending on exact CPU model.

Its a i7-4750HQ. Closing seems fine to me.
